File: solutions/delphi/collatz-conjecture/1/collatz_conjecture.pas
unit uCollatzconjecture;

interface

uses
  System.Sysutils;

type
  EIllegalNumber = class(Exception);

function collatzSteps(AValue:integer):Integer;

implementation

// SOLUTION: Recursive call
function collatzSteps(AValue:integer):Integer;
begin
  // Numbers <= 0 not allowed
  if (AValue <= 0) then
    raise EIllegalNumber.Create('error: Only positive numbers are allowed');

  // for 1 => 0 steps
  if (AValue = 1) then begin
    Result := 0;
    Exit;
  end;

  // Others numbers (use recursive call) and increment the number of steps returned
  if Odd(AValue) then
    Result := collatzSteps((Avalue * 3) + 1) + 1
  else
    Result := collatzSteps(AValue div 2) + 1;
end;

end.

File: solutions/euphoria/collatz-conjecture/1/collatz-conjecture.ex
public function steps(integer n)
    if n <= 0 then return "Only positive numbers are allowed" end if
  
    integer step_count = 0
    
    while n != 1 do
        if remainder(n, 2) = 0 then
            n = n / 2
        else
            n = n * 3 + 1
        end if
        step_count += 1
    end while
    
    return step_count
end function